README

__PROGRAM_NAME__ __PROGRAM_VERSION__

Counts for the given date yin and yang duality, classical element and
an animal in the Chinese calendar.


NEWS

This is the change log of __PROGRAM_NAME__ __PROGRAM_VERSION__

Change Log
All notable changes to this project will be documented in this file.

Unreleased

Added

Changed

Fixed

[0.0.1] - 2018-02-01

Added

Changed
- Prepared the public repository

Fixed

[0.0.1] - 2017-03-06

Added

Changed

Fixed


src_template/app/main.cpp
/*
 * This file is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
 *
 * __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "mainwindow.h"
#include <QApplication>

int main(int argc, char *argv[])
{
    QApplication a(argc, argv);
    MainWindow w;
    w.show();

    return a.exec();
}


src_template/app/calendar.h
/*
 * This file is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
 *
 * __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef CALENDAR_H
#define CALENDAR_H

#include <QDate>
#include <QDebug>

class Calendar {

    enum PartOfYear {
        P_Unknown,
        P_Small,
        P_Big
    };

    int day;
    int month;
    size_t year;
    bool isOurEra;

public:
    Calendar() {};
    ~Calendar() {};
    void setDate(int day,
                 int month,
                 size_t year,
                 bool isOurEra);

public:
    int getBasis();
    int getElement();
    int getAnimal();

private:
    PartOfYear getPart();
    QDate tableLookUp(size_t year);
};

#endif // CALENDAR_H


src_template/app/pictures.h
/*
 * This file is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
 *
 * __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef PICTURES_H
#define PICTURES_H

#include <QPixmap>
#include <vector>
#include <QDebug>

class PicturesHolder {

    std::vector<QPixmap> basises;
    std::vector<QPixmap> elements;
    std::vector<QPixmap> animals;

public:
    enum BasisIndex {
        B_None,
        B_Unknown,
        B_Yan,
        B_In
    };
    enum ElementIndex {
        E_None,
        E_Unknown,
        E_Wood,
        E_Fire,
        E_Earth,
        E_Metal,
        E_Water
    };
    enum AnimalIndex {
        A_None,
        A_Unknown,
        A_Rat,
        A_Ox,
        A_Tiger,
        A_Rabbit,
        A_Dragon,
        A_Snake,
        A_Horse,
        A_Goat,
        A_Monkey,
        A_Rooster,
        A_Dog,
        A_Pig
    };

public:
    PicturesHolder();
    ~PicturesHolder() {};

    QPixmap basis(BasisIndex index);
    QPixmap element(ElementIndex index);
    QPixmap animal(AnimalIndex index);
};

#endif // PICTURES_H


src_template/app/calendar.cpp
/*
 * This file is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
 *
 * __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "calendar.h"

void Calendar::setDate(
    int day,
    int month,
    size_t year,
    bool isOurEra)
{
    this->day = day;
    this->month = month;
    this->year = year;
    this->isOurEra = isOurEra;
}

int Calendar::getBasis()
{
    int res = -1;
    PartOfYear part = getPart();

    if (part == P_Unknown)
        return res;

    int era_after[2] = { 0, 1 };
    int era_before[2] = { 1, 0 };

    if (isOurEra) {
        if (part == P_Big) {
            res = era_after[year % 2];
        }
        else if (part == P_Small) {
            if (year > 1)
                res = era_after[(year - 1) % 2];
            else
                res = era_after[0];
        }
    } else {
        if (part == P_Big)
            res = era_before[year % 2];
        else if (part == P_Small)
            res = era_before[(year + 1) % 2];
    }
    return res;
}

int Calendar::getElement()
{
    int res = -1;
    PartOfYear part = getPart();

    if (part == P_Unknown)
        return res;

    int era_after[10] = { 3, 3, 4, 4, 0, 0, 1, 1, 2, 2 };
    int era_before[10] = { 3, 3, 2, 2, 1, 1, 0, 0, 4, 4 };

    if (isOurEra) {
        if (part == P_Big) {
            res = era_after[year % 10];
        }
        else if (part == P_Small) {
            if (year > 1)
                res = era_after[(year - 1) % 10];
            else
                res = era_after[0];
        }
    } else {
        if (part == P_Big)
            res = era_before[year % 10];
        else if (part == P_Small)
            res = era_before[(year + 1) % 10];
    }
    return res;
}

int Calendar::getAnimal()
{
    int res = -1;
    PartOfYear part = getPart();

    if (part == P_Unknown)
        return res;

    int era_after[12] = { 8, 9, 10, 11, 0, 1, 2, 3, 4, 5, 6, 7 };
    int era_before[12] = { 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 11, 10 };

    if (isOurEra) {
        if (part == P_Big) {
            res = era_after[year % 12];
        }
        else if (part == P_Small) {
            if (year > 1)
                res = era_after[(year - 1) % 12];
            else
                res = era_after[0];
        }
    } else {
        if (part == P_Big)
            res = era_before[year % 12];
        else if (part == P_Small)
            res = era_before[(year + 1) % 12];
    }
    return res;
}

Calendar::PartOfYear Calendar::getPart()
{
    // до 20 января включительно - старый год
    // после 20 февраля включительно - новый год

    if (month == 1 && day <= 20) {
        return P_Small;
    }
    if (month > 2 || (month == 2 && day >= 20)) {
        return P_Big;
    }
    if ((month == 1 && day > 20) || (month == 2 && day < 20)) {
        if (isOurEra) {
            QDate tableDate = tableLookUp(year);
            if (year == size_t(tableDate.year())) {
                QDate today(year, month, day);
                if (today >= tableDate)
                    return P_Big;
                else
                    return P_Small;
            }
        }
    }
    return P_Unknown;
}

static QDate knownDatesTable[] = {
    QDate(2016, 2, 8), QDate(1900, 1, 31), QDate(1901, 2, 19),
    QDate(1902, 2, 8), QDate(1903, 1, 29), QDate(1904, 2, 16),
    QDate(1905, 2, 4), QDate(1906, 1, 25), QDate(1907, 2, 13),
    QDate(1908, 2, 2), QDate(1909, 1, 22), QDate(1910, 2, 10),
    QDate(1911, 1, 30), QDate(1912, 2, 18), QDate(1913, 2, 6),
    QDate(1914, 1, 26), QDate(1915, 2, 14), QDate(1916, 2, 3),
    QDate(1917, 1, 23), QDate(1918, 2, 11), QDate(1919, 2, 1),
    QDate(1920, 2, 20), QDate(1921, 2, 8), QDate(1922, 1, 28),
    QDate(1923, 2, 16), QDate(1924, 2, 5), QDate(1925, 1, 25),
    QDate(1926, 2, 13), QDate(1927, 2, 2), QDate(1928, 1, 23),
    QDate(1929, 2, 10), QDate(1930, 1, 30), QDate(1931, 2, 17),
    QDate(1932, 2, 6), QDate(1933, 1, 26), QDate(1934, 2, 14),
    QDate(1935, 2, 4), QDate(1936, 1, 24), QDate(1937, 2, 11),
    QDate(1938, 1, 31), QDate(1939, 2, 19), QDate(1940, 2, 8),
    QDate(1941, 1, 27), QDate(1942, 2, 15), QDate(1943, 2, 5),
    QDate(1944, 1, 25), QDate(1945, 2, 13), QDate(1946, 2, 2),
    QDate(1947, 1, 22), QDate(1948, 2, 10), QDate(1949, 1, 29),
    QDate(1950, 2, 17), QDate(1951, 2, 6), QDate(1952, 1, 27),
    QDate(1953, 2, 14), QDate(1954, 2, 3), QDate(1955, 1, 24),
    QDate(1956, 2, 12), QDate(1957, 1, 31), QDate(1958, 2, 18),
    QDate(1959, 2, 8), QDate(1960, 1, 28), QDate(1961, 2, 15),
    QDate(1962, 2, 5), QDate(1963, 1, 25), QDate(1964, 2, 13),
    QDate(1965, 2, 2), QDate(1966, 1, 21), QDate(1967, 2, 9),
    QDate(1968, 1, 30), QDate(1969, 2, 17), QDate(1970, 2, 6),
    QDate(1971, 1, 27), QDate(1972, 2, 15), QDate(1973, 2, 3),
    QDate(1974, 1, 23), QDate(1975, 2, 11), QDate(1976, 1, 31),
    QDate(1977, 2, 18), QDate(1978, 2, 7), QDate(1979, 1, 28),
    QDate(1980, 2, 16), QDate(1981, 2, 5), QDate(1982, 1, 25),
    QDate(1983, 2, 13), QDate(1984, 2, 2), QDate(1985, 2, 20),
    QDate(1986, 2, 9), QDate(1987, 1, 29), QDate(1988, 2, 17),
    QDate(1989, 2, 6), QDate(1990, 1, 27), QDate(1991, 2, 15),
    QDate(1992, 2, 4), QDate(1993, 1, 23), QDate(1994, 2, 10),
    QDate(1995, 1, 31), QDate(1996, 2, 19), QDate(1997, 2, 7),
    QDate(1998, 1, 28), QDate(1999, 2, 16), QDate(2000, 2, 5),
    QDate(2001, 1, 24), QDate(2002, 2, 12), QDate(2003, 2, 1),
    QDate(2004, 1, 22), QDate(2005, 2, 9), QDate(2006, 1, 29),
    QDate(2007, 2, 18), QDate(2008, 2, 7), QDate(2009, 1, 26),
    QDate(2010, 2, 14), QDate(2011, 2, 3), QDate(2012, 1, 23),
    QDate(2013, 2, 10), QDate(2014, 1, 31), QDate(2015, 2, 19),
    QDate(2016, 2, 8), QDate(2017, 1, 28), QDate(2018, 2, 16),
    QDate(2019, 2, 5), QDate(2020, 1, 25), QDate(2021, 2, 12),
    QDate(2022, 2, 1), QDate(2023, 1, 22), QDate(2024, 2, 10),
    QDate(2025, 1, 29), QDate(2026, 2, 17), QDate(2027, 2, 6),
    QDate(2028, 1, 26), QDate(2029, 2, 13), QDate(2030, 2, 3),
    QDate(2031, 1, 23), QDate(2032, 2, 11), QDate(2033, 1, 31),
    QDate(2034, 2, 19), QDate(2035, 2, 8), QDate(2036, 1, 28),
    QDate(2037, 2, 15), QDate(2038, 2, 4), QDate(2039, 2, 24),
    QDate(2040, 2, 12), QDate(2041, 2, 1), QDate(2042, 2, 22),
    QDate(2043, 2, 10),
};

QDate Calendar::tableLookUp(size_t year)
{
    size_t size =
        sizeof knownDatesTable / sizeof knownDatesTable[0];
    for (size_t i = 0; i < size; i++) {
        if (size_t(knownDatesTable[i].year()) == year)
            return knownDatesTable[i];
    }
    return QDate(0, 0, 0);
}


src_template/app/mainwindow.h
/*
 * This file is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
 *
 * __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include <QMainWindow>
#include <QDesktopWidget>
#include <QDate>
#include <QMessageBox>
#include <QKeyEvent>
#include <QSizePolicy>
#include <QDebug>
#include "pictures.h"
#include "calendar.h"

namespace Ui {
class MainWindow;
}

class MainWindow : public QMainWindow
{
    Q_OBJECT

public:
    explicit MainWindow(QWidget *parent = 0);
    ~MainWindow();
   void moveToCenter();
   void keyPressEvent(QKeyEvent *event);
   bool eventFilter(QObject *object, QEvent *event);

private slots:
    void on_pushButton_eval_clicked();

    void on_pushButton_clear_clicked();

    void on_action_triggered();

    void slot1_alignNumberOfDays();

    void slot2_decreaseYear();

    void slot3_increaseYear();

private:
    Ui::MainWindow *ui;
    PicturesHolder *pictures;
    Calendar *calendar;
    bool focusOnRadio;

public:
    void clearDate();
    void clearResult();

private:
    bool isLeapYear(size_t year);
    bool isLeapNegativeYear(size_t year);
};

#endif // MAINWINDOW_H


src_template/app/pictures.cpp
/*
 * This file is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
 *
 * __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "pictures.h"

PicturesHolder::PicturesHolder()
{
    QString path;

    path = QString::fromUtf8(":/images/basis/images/basis/basis-none.png");
    basises.push_back(QPixmap(path));

    path = QString::fromUtf8(":/images/basis/images/basis/basis-unknown.png");
    basises.push_back(QPixmap(path));

    path = QString::fromUtf8(":/images/basis/images/basis/basis-yan.png");
    basises.push_back(QPixmap(path));

    path = QString::fromUtf8(":/images/basis/images/basis/basis-in.png");
    basises.push_back(QPixmap(path));


    path = QString::fromUtf8(
        ":/images/elements/images/elements/elements-none.png");
    elements.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/elements/images/elements/elements-unknown.png");
    elements.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/elements/images/elements/elements-wood1.png");
    elements.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/elements/images/elements/elements-fire2.png");
    elements.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/elements/images/elements/elements-earth3.png");
    elements.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/elements/images/elements/elements-metal4.png");
    elements.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/elements/images/elements/elements-water5.png");
    elements.push_back(QPixmap(path));


    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-none.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-unknown.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-rat1.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-ox2.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-tiger3.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-rabbit4.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-dragon5.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-snake6.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-horse7.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-goat8.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-monkey9.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-rooster10.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-dog11.png");
    animals.push_back(QPixmap(path));

    path = QString::fromUtf8(
        ":/images/animals/images/animals/animals-pig12.png");
    animals.push_back(QPixmap(path));
}

QPixmap PicturesHolder::basis(BasisIndex index)
{
    return basises[index];
}

QPixmap PicturesHolder::element(ElementIndex index)
{
    return elements[index];
}

QPixmap PicturesHolder::animal(AnimalIndex index)
{
    return animals[index];
}


src_template/app/mainwindow.cpp
/*
 * This file is a part of __PROGRAM_NAME__ __PROGRAM_VERSION__
 *
 * __PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__ __PROGRAM_AUTHOR_EMAIL__
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "mainwindow.h"
#include "ui_mainwindow.h"

MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    moveToCenter();

    pictures = new PicturesHolder();
    calendar = new Calendar();

    focusOnRadio = false;
    ui->radioButton_era->installEventFilter(this);
    ui->radioButton_era_before->installEventFilter(this);

    clearDate();
    clearResult();
}

MainWindow::~MainWindow()
{
    delete calendar;
    delete pictures;
    delete ui;
}

void MainWindow::moveToCenter()
{
    QRect rect = frameGeometry();
    rect.moveCenter(QDesktopWidget().availableGeometry().center());
    move(rect.topLeft());
}

void MainWindow::keyPressEvent(QKeyEvent *event)
{
    switch (event->key()) {
    case Qt::Key_Left:
        if (!focusOnRadio)
            slot2_decreaseYear();
        break;
    case Qt::Key_Right:
        if (!focusOnRadio)
            slot3_increaseYear();
        break;
    case Qt::Key_Return:
        on_pushButton_eval_clicked();
        break;
    case Qt::Key_Escape:
        close();
        break;
    default:
        QWidget::keyPressEvent(event);
        return;
    }
}

void MainWindow::clearDate()
{
    QDate currentDate = QDate::currentDate();
    int day = currentDate.day();
    int month = currentDate.month();
    size_t year = currentDate.year();

    ui->comboBox_month->setCurrentIndex(month - 1);
    ui->spinBox_day->setValue(day);
    ui->spinBox_year->setValue(year);
    ui->radioButton_era->setChecked(true);
}

void MainWindow::clearResult()
{
    QPixmap basis = pictures->basis(PicturesHolder::B_None);
    QPixmap element = pictures->element(PicturesHolder::E_None);
    QPixmap animal = pictures->animal(PicturesHolder::A_None);
    ui->label_basis_img->setPixmap(basis);
    ui->label_element_img->setPixmap(element);
    ui->label_animal_img->setPixmap(animal);
}

void MainWindow::on_pushButton_eval_clicked()
{
    int day, month;
    size_t year;
    bool isOurEra;

    year = ui->spinBox_year->value();
    month = ui->comboBox_month->currentIndex() + 1;
    day = ui->spinBox_day->value();
    isOurEra = ui->radioButton_era->isChecked();

    calendar->setDate(day, month, year, isOurEra);

    QPixmap basisImg, elementImg, animalImg;

    switch (calendar->getBasis()) {
    case 0:
        basisImg = pictures->basis(PicturesHolder::B_Yan);
        break;
    case 1:
        basisImg = pictures->basis(PicturesHolder::B_In);
        break;
    case -1:
        basisImg = pictures->basis(PicturesHolder::B_Unknown);
        break;
    }

    switch (calendar->getElement()) {
    case 0:
        elementImg = pictures->element(PicturesHolder::E_Wood);
        break;
    case 1:
        elementImg = pictures->element(PicturesHolder::E_Fire);
        break;
    case 2:
        elementImg = pictures->element(PicturesHolder::E_Earth);
        break;
    case 3:
        elementImg = pictures->element(PicturesHolder::E_Metal);
        break;
    case 4:
        elementImg = pictures->element(PicturesHolder::E_Water);
        break;
    case -1:
        elementImg = pictures->element(PicturesHolder::E_Unknown);
        break;
    }

    switch (calendar->getAnimal()) {
    case 0:
        animalImg = pictures->animal(PicturesHolder::A_Rat);
        break;
    case 1:
        animalImg = pictures->animal(PicturesHolder::A_Ox);
        break;
    case 2:
        animalImg = pictures->animal(PicturesHolder::A_Tiger);
        break;
    case 3:
        animalImg = pictures->animal(PicturesHolder::A_Rabbit);
        break;
    case 4:
        animalImg = pictures->animal(PicturesHolder::A_Dragon);
        break;
    case 5:
        animalImg = pictures->animal(PicturesHolder::A_Snake);
        break;
    case 6:
        animalImg = pictures->animal(PicturesHolder::A_Horse);
        break;
    case 7:
        animalImg = pictures->animal(PicturesHolder::A_Goat);
        break;
    case 8:
        animalImg = pictures->animal(PicturesHolder::A_Monkey);
        break;
    case 9:
        animalImg = pictures->animal(PicturesHolder::A_Rooster);
        break;
    case 10:
        animalImg = pictures->animal(PicturesHolder::A_Dog);
        break;
    case 11:
        animalImg = pictures->animal(PicturesHolder::A_Pig);
        break;
    case -1:
        animalImg = pictures->animal(PicturesHolder::A_Unknown);
        break;
    }

    ui->label_basis_img->setPixmap(basisImg);
    ui->label_element_img->setPixmap(elementImg);
    ui->label_animal_img->setPixmap(animalImg);
}

void MainWindow::on_pushButton_clear_clicked()
{
    clearDate();
    clearResult();
}

void MainWindow::on_action_triggered()
{
    QString title = QString::fromUtf8("О программе");
    QString msg = QString::fromUtf8(
        "__PROGRAM_NAME__ __PROGRAM_VERSION__\n"
        "\n"
        "Вычислияет начало, стихию и животное по восточному календарю.\n"
        "\n"
        "__PROGRAM_COPYRIGHT__ __PROGRAM_AUTHOR__"
        "\n"
        "__PROGRAM_AUTHOR_EMAIL__"
    );

    QMessageBox::information(this, title, msg);
}

void MainWindow::slot1_alignNumberOfDays()
{
    int monthDays[] = {31, 28, 31, 30,
                       31, 30, 31, 31,
                       30, 31, 30, 31};
    int month = ui->comboBox_month->currentIndex() + 1;
    size_t year = ui->spinBox_year->value();
    bool isOurEra = ui->radioButton_era->isChecked();

    if (month == 2) {
        int value = monthDays[month - 1];
        if ((isOurEra && isLeapYear(year)) ||
            (!isOurEra && isLeapNegativeYear(year)))
            value++;
        ui->spinBox_day->setMaximum(value);
    } else {
        ui->spinBox_day->setMaximum(monthDays[month - 1]);
    }
}

bool MainWindow::eventFilter(QObject *object, QEvent *event)
{
    if (event->type() == QEvent::FocusIn) {
        if (object == ui->radioButton_era ||
            object == ui->radioButton_era_before) {
            focusOnRadio = true;
        }
    }
    if (event->type() == QEvent::FocusOut) {
        if (object == ui->radioButton_era ||
            object == ui->radioButton_era_before) {
            focusOnRadio = false;
        }
    }
    return false;
}

void MainWindow::slot2_decreaseYear()
{
    size_t year = ui->spinBox_year->value();
    bool isOurEra = ui->radioButton_era->isChecked();

    if (isOurEra) {
        if (year > 1)
            ui->spinBox_year->setValue(year - 1);
        else
            ui->radioButton_era_before->setChecked(true);
    } else {
        ui->spinBox_year->setValue(year + 1);
    }
}

void MainWindow::slot3_increaseYear()
{
    size_t year = ui->spinBox_year->value();
    bool isOurEra = ui->radioButton_era->isChecked();

    if (isOurEra) {
        ui->spinBox_year->setValue(year + 1);
    } else {
        if (year > 1)
            ui->spinBox_year->setValue(year - 1);
        else
            ui->radioButton_era->setChecked(true);
    }
}

bool MainWindow::isLeapYear(size_t year)
{
    // 500 делится на 100 и на 4, но не делится на 400
    // 400 делится на 100 и на 4, и делится на 400
    return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
}

bool MainWindow::isLeapNegativeYear(size_t year)
{
    // 1 год до нашей эры - это високосный год столетия
    // до нашей эры 1, 5, 9 ... - это годы через 4
    // до нашей эры 1, 401, 801, ... - это годы столетий
    year += 399;
    return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
}
